cli/tools/test.rs

```diff
@@ -153,13 +158,100 @@ pub enum TestOutput {
   Bytes(Vec<u8>),
 }
 
+#[allow(clippy::derive_partial_eq_without_eq)]
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum TestFailure {
+  JsError(Box<JsError>),
+  FailedSteps(usize),
+  IncompleteSteps,
+  LeakedOps(Vec<String>, bool), // Details, isOpCallTracingEnabled
+  LeakedResources(Vec<String>), // Details
+  // The rest are for steps only.
+  Incomplete,
+  OverlapsWithSanitizers(IndexSet<String>), // Long names of overlapped tests
+  HasSanitizersAndOverlaps(IndexSet<String>), // Long names of overlapped tests
+}
+
+impl ToString for TestFailure {
+  fn to_string(&self) -> String {
+    match self {
+      TestFailure::JsError(js_error) => format_test_error(js_error),
+      TestFailure::FailedSteps(1) => "1 test step failed.".to_string(),
+      TestFailure::FailedSteps(n) => format!("{} test steps failed.", n),
+      TestFailure::IncompleteSteps => "Completed while steps were still running. Ensure all steps are awaited with `await t.step(...)`.".to_string(),
+      TestFailure::Incomplete => "Didn't complete before parent. Await step with `await t.step(...)`.".to_string(),
+      TestFailure::LeakedOps(details, is_op_call_tracing_enabled) => {
+        let mut string = "Leaking async ops:".to_string();
+        for detail in details {
+          string.push_str(&format!("\n  - {}", detail));
+        }
+        if !is_op_call_tracing_enabled {
+          string.push_str("\nTo get more details where ops were leaked, run again with --trace-ops flag.");
+        }
+        string
+      }
+      TestFailure::LeakedResources(details) => {
+        let mut string = "Leaking resources:".to_string();
+        for detail in details {
+          string.push_str(&format!("\n  - {}", detail));
+        }
+        string
+      }
+      TestFailure::OverlapsWithSanitizers(long_names) => {
+        let mut string = "Started test step while another test step with sanitizers was running:".to_string();
+        for long_name in long_names {
+          string.push_str(&format!("\n  * {}", long_name));
+        }
+        string
+      }
+      TestFailure::HasSanitizersAndOverlaps(long_names) => {
+        let mut string = "Started test step with sanitizers while another test step was running:".to_string();
```

Comment:
**nitpick:** If I'm not badly mistaken, there's some way to join a vector of strings or possibly concatenate a string iterator, so it should be possible to do these error message creations without needing to use `push_str`. Not sure if it would be any better in any shape or form, though :)

Reply:
I imagined that looking like `format!(..., foo.iter().map(|s| format!(..., s)).intersperse("\n").collect())` and went with this way. Would be nice if a macro existed like
```rs
concat2!(
  "Leaking async ops:\n",
  "  - {details[#]}\n",
  "To get more details where ops were leaked, run again with --trace-ops flag."
)
```